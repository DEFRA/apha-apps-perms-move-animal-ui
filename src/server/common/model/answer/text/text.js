import Joi from 'joi'
import { AnswerModel } from '../answer-model.js'
import { validateAnswerAgainstSchema } from '../validation.js'
import { NotImplementedError } from '../../../helpers/not-implemented-error.js'
import { escapeHtml } from '../../../helpers/html-escape.js'

/** @import {AnswerViewModelOptions} from '../answer-model.js' */

/**
 * @param {TextConfig} config
 * @returns {Joi.Schema}
 */
const textSchema = ({ payloadKey, validation, stripWhitespace }) => {
  let stringValidation = Joi.string().trim().required()

  const messages = {
    'any.required': validation.empty?.message ?? '',
    'string.empty': validation.empty?.message ?? '',
    'string.max': validation.maxLength?.message ?? ''
  }

  if (validation.maxLength) {
    stringValidation = stringValidation.max(validation.maxLength.value)
  }

  if (validation.pattern) {
    stringValidation = stringValidation.pattern(validation.pattern.regex)
    messages['string.pattern.base'] = validation.pattern.message
  }

  if (stripWhitespace) {
    stringValidation = stringValidation.replace(whitespaceRegex, '')
  }

  return Joi.object({
    [payloadKey]: stringValidation.messages(messages)
  })
}

const whitespaceRegex = /\s+/g

/**
 * export @typedef {{
 *  payloadKey: string,
 *  stripWhitespace?: boolean,
 *  type?: 'email',               // the default is text, so no need to specify
 *  spellcheck?: false,
 *  characterWidth?: 2 | 10 | 20,
 *  isPageHeading? : boolean,
 *  autocomplete?: string,
 *  hint?: string,
 *  validation: {
 *    maxLength?: { value: number, message: string },
 *    empty?: { message: string },
 *    pattern?: { regex: RegExp, message: string }
 *  }
 * }} TextConfig
 */

/**
 * @typedef {string} TextData
 */

/**
 * @template Payload
 * @augments {AnswerModel<Payload>}
 */
export class TextAnswer extends AnswerModel {
  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {TextConfig} */
  get config() {
    return /** @type {any} */ (this.constructor).config
  }

  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {TextConfig} */
  static get config() {
    throw new NotImplementedError()
  }

  /**
   * @returns {string | undefined}
   */
  get value() {
    return this._data?.[this.config.payloadKey]
  }

  get html() {
    return escapeHtml(this._data?.[this.config.payloadKey] ?? '')
  }

  // eslint-disable-next-line @typescript-eslint/class-literal-property-style
  get template() {
    return 'model/answer/text/text.njk'
  }

  /**
   * @returns {TextData}
   */
  toState() {
    let v = this.value?.trim() ?? ''

    if (this.config.stripWhitespace) {
      v = v.replace(whitespaceRegex, '')
    }
    return v
  }

  validate() {
    return validateAnswerAgainstSchema(
      textSchema(this.config),
      this._data ?? {}
    )
  }

  /**
   * @param {AnswerViewModelOptions} options
   */
  viewModel({ validate, question }) {
    const { payloadKey, type, spellcheck, autocomplete, characterWidth, hint } =
      this.config

    const isPageHeading = this.config.isPageHeading ?? true

    const viewModel = {
      label: {
        text: question,
        classes: isPageHeading ? 'govuk-label--l' : 'govuk-label--m',
        isPageHeading
      },
      id: payloadKey,
      name: payloadKey,
      value: this.value
    }

    if (characterWidth) {
      viewModel.classes = `govuk-input--width-${characterWidth}`
    }

    if (autocomplete) {
      viewModel.autocomplete = autocomplete
    }

    if (type) {
      viewModel.type = type
    }

    if (hint) {
      viewModel.hint = { text: hint }
    }

    if (spellcheck === false) {
      viewModel.spellcheck = false
    }

    if (validate) {
      viewModel.errorMessage = this.validate().errors[payloadKey]
    }

    return viewModel
  }

  /**
   * @param {Payload} fields
   */
  _extractFields(fields) {
    return /** @type {Payload} */ ({
      [this.config.payloadKey]: fields[this.config.payloadKey]
    })
  }

  /**
   * @param {TextData | undefined} state
   * @returns {TextAnswer}
   */
  static fromState(state) {
    return new this(
      state !== undefined ? { [this.config.payloadKey]: state } : {}
    )
  }
}
