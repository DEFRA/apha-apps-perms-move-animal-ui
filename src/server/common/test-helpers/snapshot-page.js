import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
