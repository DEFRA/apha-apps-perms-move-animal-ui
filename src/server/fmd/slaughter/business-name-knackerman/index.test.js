import { describePageSnapshot } from '~/src/server/common/test-helpers/snapshot-page.js'
import { Answer, businessNameKnackermanPage } from './index.js'
import { TextAnswer } from '~/src/server/common/model/answer/text/text.js'
import { NumberKnackermanPage } from '../number-knackerman/index.js'

const sectionKey = 'slaughter'
const questionKey = 'businessNameKnackerman'
const pageUrl = '/fmd/slaughter-information/knackerman-business-name'
const page = businessNameKnackermanPage
const question = 'What is the business name of the Knackerman?'

const payload = {
  [questionKey]: 'some text'
}

describe('Answer', () => {
  it('should be a Text input', () => {
    expect(new Answer(payload)).toBeInstanceOf(TextAnswer)
  })

  it('should have the right payload key', () => {
    expect(Answer.config.payloadKey).toBe(questionKey)
  })

  it('should have the correct character width', () => {
    expect(Answer.config.characterWidth).toBe(10)
  })

  it('should have spellcheck disabled', () => {
    expect(Answer.config.spellcheck).toBe(false)
  })

  it('should have the correct validation rules', () => {
    expect(Answer.config.validation).toBeDefined()
    expect(Answer.config.validation?.empty?.message).toBe(
      "Enter the Knackerman's business name"
    )
    expect(Answer.config.validation?.maxLength?.value).toBe(100)
    expect(Answer.config.validation?.maxLength?.message).toBe(
      'Your answer must be no longer than 100 characters'
    )
  })
})

describe('BusinessNameKnackermanPage', () => {
  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(Answer)
  })

  describe('nextPage', () => {
    it('should return NumberKnackermanPage for any value', () => {
      const nextPage = page.nextPage()
      expect(nextPage).toBeInstanceOf(NumberKnackermanPage)
    })
  })

  describePageSnapshot({
    describes: 'content',
    it: 'should render expected response and content',
    pageUrl
  })
})
