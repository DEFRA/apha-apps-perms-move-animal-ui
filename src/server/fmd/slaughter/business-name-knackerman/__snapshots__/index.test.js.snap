// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusinessNameKnackermanPage content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<fieldset class="govuk-fieldset">
  <div class="govuk-form-group">
  <h1 class="govuk-label-wrapper">
    <label class="govuk-label govuk-label--l" for="businessNameKnackerman">
      What is the business name of the Knackerman?
    </label>
  </h1>
  <input class="govuk-input govuk-input--width-10" id="businessNameKnackerman" name="businessNameKnackerman" type="text" spellcheck="false">
</div>


</fieldset>
            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
